### PR TITLE
Timer PR parts 1 and 2 from #146

### PR DESCRIPTION
**Pull request by ravi-mosaicml**
Added the `Time` class, `Timer` class, conversion function, and test cases and docstrings for these classes. This functionality encompasses parts 1 and 2 from #146, with the following changes:

1. `tokens` will be 0 (instead of None) if not being tracked.

Also, modified the `StringEnum` class to relax the requirement that the lowercase name be the value, since this does not hold for our timing abbreviations.

Other PRs will integrate the Timer into the training loop.

### CODE REVIEW

File: composer/core/time.py [hunk 1]
```diff
@@ -0,0 +1,505 @@
...
+from typing import TYPE_CHECKING, Optional, Tuple, Union
+
+from composer.core.serializable import Serializable
+from composer.utils.string_enum import StringEnum
+
+if TYPE_CHECKING:
+    from composer.core.types import StateDict
+
+
+class TimeUnit(StringEnum):
```
Comment (hanlint) [inline composer/core/time.py:17]: does `TimeUnit(str, Enum)` work here?

**Review by hanlint — CHANGES_REQUESTED** [review]
I like where this is headed! I made a few suggestions around simplifying the user burden.

File: composer/core/time.py (hunk 1 above)
Comment (ravi-mosaicml) [reply composer/core/time.py:17]: The `StringEnum` class provides for additional functionality, such as being case-insensitive and for allowing comparisons between strings and the enum class. Would prefer to use it here as we use it elsewhere too.

File: composer/core/time.py [hunk 2]
```diff
@@ -0,0 +1,505 @@
...
+
+            >>> Time(5, TimeUnit.EPOCH)  # describes 5 epochs.
+            >>> Time(3e4, "tok")  # describes 30,000 tokens.
+            >>> Time(0.5, "dur")  # describes 50% of the training process.
+        
+        #. Use a keyword argument for the unit. For example,
+
+            >>> Time(epoch=5)  # describes 5 epochs.
+            >>> Time(token=3e4)  # describes 30,000 tokens.
+            >>> Time(dur=0.5)  # describes 50% of the training process.
```
[the remark this answers is not in the document]
Comment (ravi-mosaicml) [reply composer/core/time.py:65]: Whoops, this is a bug in the docs. But this will be a nonissue with https://github.com/mosaicml/composer/pull/174/files#r771913417.

File: composer/core/time.py [hunk 3]
```diff
@@ -0,0 +1,505 @@
...
+        match = [x for x in match if x != '']
+        assert len(match) == 2, "each match should have a number followed by the key"
+        value = match[0]
+        unit = TimeUnit(match[1])
+        value = float(value)  # always parsing first as float b/c it could be scientific notation
+        if unit != TimeUnit.DURATION:
+            value = int(value)
+        return value, unit
+
+    def convert(
```
[the remark this answers is not in the document]
Comment (ravi-mosaicml) [reply composer/core/time.py:269]: I mocked this out for `to_epoch` an mostly every parameter would still have to be optional -- it depends a lot on what the from unit is. I can add in a matrix into the docs to make it clearer.

File: composer/core/time.py [hunk 4]
```diff
@@ -0,0 +1,505 @@
...
+        return self._unit
+
+    def __repr__(self) -> str:
+        return f"{self.__class__.__name__}({self.unit.name.lower()}={self.value})"
+
+    def _convert(self, other: object) -> Time:
+        if isinstance(other, Time):
+            return other
+
+        if isinstance(other, (int, float)):
```
[the remark this answers is not in the document]
Comment (ravi-mosaicml) [reply composer/core/time.py:160]: Disabled it for comparisons, addition, and subtraction. Left it for multiplication and division.

File: composer/core/time.py [hunk 5]
```diff
@@ -0,0 +1,505 @@
...
+            >>> Time(token=3e4)  # describes 30,000 tokens.
+            >>> Time(dur=0.5)  # describes 50% of the training process.
+
+    :class:`Time` supports addition and subtraction with other :class:`Time` instances that share the same
+    :class:`TimeUnit`. For example:
+    >>> Time(epoch=1) + Time(epoch=2) == Time("3ep")
+
+    :class:`Time` supports multiplication when the multiplier or multiplicand is in units :attr:`TimeUnit.DURATION`.
+    For example:
+    >>> Time(epoch=2) * Time(0.5, "dur) == Time("1ep")
```
[the remark this answers is not in the document]
Comment (ravi-mosaicml) [reply composer/core/time.py]: This is supported; added a test case to confirm.

File: composer/core/time.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (ravi-mosaicml) [reply composer/core/time.py:303]: I added in helper functions per @Averylamp's suggestion

[the remark this answers is not in the document]
Comment (abhi-mosaic) [reply composer/core/time.py:303]: So, looking back at #146 and the ways we want `Time` and `Timer` to be used, I think `Time` conversions should be considered super dangerous, and should maybe even log a warning. They cannot be guaranteed to be correct either forwards or backwards in time, because we could be dynamically editing batch size, seq_len, filtering out data from the dataset, etc. 

Ideally, the only object that should ever perform `Time` conversions is the `Trainer` upon initialization, to convert milestones to a shared base `TimeUnit` and even then I'm not sure if it is actually needed... see this example YAML (https://github.com/mosaicml/composer/issues/146#issuecomment-994031037). The key question is "how do I specify that all durations relative to `90ep` should be measured in samples `sp`? Because I need the warmup period to be 5% of `90ep`, but granular down to the sample."

To this end, I would prefer if `Time` conversion was only something that a `Timer` or `Trainer` object could do but not attached to `Time`.

One related issue with naming, the names like `dataset_num_samples`, `batch_size` is all very `Trainer` specific, what is really needed for conversion is `num_samples_per_epoch`, `samples_per_batch`, `tokens_per_batch`... The former is sort of naming what `Trainer` would pass as the values to the convert function, which requires knowing how dataloaders + DDP work

[the remark this answers is not in the document]
Comment (abhi-mosaic) [reply composer/core/time.py:303]: One spec I can imagine is, `Time` units have no conversions, and the `Timer` is given a `base_unit` which tells it the granularity with which to increment its internal `timer._duration`. This way, any milestones specified like `0.1dur` will be able to compare with `Timer`'s duration and figure out when to trigger.

Also, `Timer` may need to have both global and period counters, since we will likely want to say things like "`0.1dur` of the post-warmup period".

[the remark this answers is not in the document]
Comment (ravi-mosaicml) [reply composer/core/time.py:303]: That makes sense -- I don't think users should need to call convert. I can make it private. In that case, since it will be private, is it OK to have the one function with a bunch of optional parameters?

[the remark this answers is not in the document]
Comment (ravi-mosaicml) [reply composer/core/time.py:303]: Re: duration of training periods, would it be OK if we did this as a separate feature? It would be helpful for me if we had a more detailed design proposal for this (e.g. how would users specify the length of periods? Can you mix and match units for period duration? How many periods would there be? Can periods be user-defined in the yaml?) I don't think it would be too hard to extend the timing object to support a duration relative to a period, but would prefer to get this core functionality merged in first.

**Review by hanlint — APPROVED** [review]
